core/gesture.js

```diff
@@ -406,39 +448,31 @@ Blockly.Gesture.prototype.handleUp = function(e) {
   } else if (this.isWorkspaceClick_()) {
     this.doWorkspaceClick_();
   }
-  this.endGesture(e);
+
+  e.preventDefault();
+  e.stopPropagation();
+
+  this.dispose();
 };
 
 /**
  * Cancel an in-progress gesture.  If a workspace or block drag is in progress,
  * end the drag at the most recent location.
- *
  */
-Blockly.Gesture.prototype.cancelGesture = function() {
+Blockly.Gesture.prototype.cancel = function() {
+  // Disposing of a block cancels in-progress drags, but dragging to a delete
+  // area disposes of a block and leads to recursive disposal. Break that cycle.
+  if (this.isEnding_) {
```

Comment:
I didn't name it isDeleting because the block dragger knows whether we're deleting a block, but the owning gesture doesn't.  

The owning gesture knows that it's ending and entering a cleanup stage (happening from an event); cancel could be called from a different event midway through cleanup, leading to multiple dispose calls, or from block.dispose.